perf/perf.go

```diff
@@ -0,0 +1,543 @@
+	for i := 0; i < nCPU; i++ {
+		ring, err := newPerfEventRing(i, opts)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to create perf ring for CPU %d", i)
+		}
+		rings = append(rings, ring)
+
+		if err := opts.Map.Put(uint32(i), uint32(ring.fd)); err != nil {
+			return nil, errors.Wrapf(err, "could't put event fd for CPU %d", i)
+		}
+
+		if err := addToEpoll(epollFd, ring.fd, len(rings)-1); err != nil {
+			return nil, err
+		}
+	}
+
+	closeFd, err := unix.Eventfd(0, unix.O_CLOEXEC|unix.O_NONBLOCK)
+	if err != nil {
+		return nil, err
+	}
+	fds = append(fds, closeFd)
+
+	if err := addToEpoll(epollFd, closeFd, -1); err != nil {
+		return nil, err
+	}
+
+	array, err := opts.Map.Clone()
+	if err != nil {
+		return nil, err
+	}
+
+	pr = &Reader{
+		array:   array,
+		rings:   rings,
+		epollFd: epollFd,
+		// Allocate extra event for closeFd
+		epollEvents: make([]unix.EpollEvent, len(rings)+1),
+		epollRings:  make([]*perfEventRing, 0, len(rings)),
+		closeFd:     closeFd,
+	}
+	runtime.SetFinalizer(pr, (*Reader).Close)
+	return pr, nil
+}
+
+// Close frees resources used by the reader.
+//
+// It interrupts calls to Read.
+//
+// Calls to perf_event_output from eBPF programs will return
+// ENOENT after calling this method.
+func (pr *Reader) Close() error {
+	var err error
+	pr.closeOnce.Do(func() {
+		runtime.SetFinalizer(pr, nil)
+
+		// Interrupt Read() via the event fd.
+		var value [8]byte
+		internal.NativeEndian.PutUint64(value[:], 1)
+		_, err = unix.Write(pr.closeFd, value[:])
+		if err != nil {
+			err = errors.Wrap(err, "can't write event fd")
+			return
+		}
+
+		// Acquire the lock. This ensures that Read
+		// isn't running.
+		pr.mu.Lock()
+		defer pr.mu.Unlock()
+
+		unix.Close(pr.epollFd)
+		unix.Close(pr.closeFd)
+		pr.epollFd, pr.closeFd = -1, -1
+
+		// Close rings
+		for _, ring := range pr.rings {
+			ring.Close()
+		}
+		pr.rings = nil
+
+		pr.array.Close()
+	})
+
+	return errors.Wrap(err, "close PerfReader")
+}
+
+// Read the next record from the perf ring buffer.
+//
+// The function blocks until there are at least Watermark bytes in one
+// of the per CPU buffers.
+//
+// Records from buffers below the Watermark are not returned.
+//
+// Calling Close interrupts the function.
+func (pr *Reader) Read() (*Record, error) {
+	var rec Record
+	err := pr.ReadRecord(&rec)
+	if err != nil {
+		return nil, err
+	}
+	return &rec, nil
+}
+
+// ReadRecord reads into a pre-allocated Record.
+//
+// Record.RawSample will be re-allocated if the buffer is too
+// small to hold the full sample.
+//
+// See Read for further information.
+func (pr *Reader) ReadRecord(rec *Record) error {
+	pr.mu.Lock()
+	defer pr.mu.Unlock()
+
+	if pr.epollFd == -1 {
+		return errClosed
+	}
+
+	for {
+		if len(pr.epollRings) == 0 {
+			nEvents, err := unix.EpollWait(pr.epollFd, pr.epollEvents, -1)
+
+			// Handle EINTR
+			if temp, ok := err.(temporaryError); ok && temp.Temporary() {
+				continue
```

Comment:
Why does this need to continue, rather than breaking the loop?

Reply:
Because we need to retry the syscall, see https://github.com/golang/go/issues/20400

Reply:
Ah, so the interruption here is intended to be ignored and we just continue with the dump?

I think a comment with the intent and a link to the above golang thread would be useful in the code.

Reply:
Done!